omniledger/service/contracts.go

```diff
@@ -163,35 +168,80 @@ func (s *Service) ContractConfig(cdb CollectionView, tx Instruction, coins []Coi
 				DarcID:     tx.ObjectID.DarcID,
 				InstanceID: OneNonce,
 			}, ContractConfigID, configBuf),
-	}, nil, nil
+	}, c, nil
 }
 
 // ContractDarc accepts the following instructions:
 //   - Spawn - creates a new darc
 //   - Invoke.Evolve - evolves an existing darc
 func (s *Service) ContractDarc(coll CollectionView, tx Instruction,
 	coins []Coin) ([]StateChange, []Coin, error) {
-	if tx.getType() != invokeType {
-		return nil, nil, errors.New("Darc can only be invoked (evolved)")
-	}
-	if tx.Invoke.Command == "evolve" {
-		darcBuf := tx.Invoke.Args.Search("darc")
-		newD, err := darc.NewDarcFromProto(darcBuf)
-		if err != nil {
-			return nil, nil, err
+	switch {
+	case tx.Spawn != nil:
+		if tx.Spawn.ContractID == ContractDarcID {
+			darcBuf := tx.Spawn.Args.Search("darc")
+			d, err := darc.NewDarcFromProto(darcBuf)
+			if err != nil {
+				return nil, nil, errors.New("given darc was not valid: " + err.Error())
+			}
+			return []StateChange{
+				NewStateChange(Create, ObjectID{d.GetBaseID(), ZeroNonce}, ContractDarcID, darcBuf),
+			}, coins, nil
 		}
-		oldD, err := LoadDarcFromColl(coll, ObjectID{newD.BaseID, ZeroNonce}.Slice())
-		if err != nil {
-			return nil, nil, err
+		c, found := s.contracts[tx.Spawn.ContractID]
+		if !found {
+			return nil, nil, errors.New("couldn't find this contract type")
 		}
-		if err := newD.SanityCheck(oldD); err != nil {
-			return nil, nil, err
+		return c(coll, tx, coins)
+	case tx.Invoke != nil:
+		switch tx.Invoke.Command {
+		case "evolve":
+			darcBuf := tx.Invoke.Args.Search("darc")
+			newD, err := darc.NewDarcFromProto(darcBuf)
+			if err != nil {
+				return nil, nil, err
+			}
+			oldD, err := LoadDarcFromColl(coll, ObjectID{newD.BaseID, ZeroNonce}.Slice())
+			if err != nil {
+				return nil, nil, err
+			}
+			if err := newD.SanityCheck(oldD); err != nil {
+				return nil, nil, err
+			}
+			return []StateChange{
+				NewStateChange(Update, tx.ObjectID, ContractDarcID, darcBuf),
+			}, coins, nil
+		default:
+			return nil, nil, errors.New("invalid command: " + tx.Invoke.Command)
 		}
-		return []StateChange{
-			NewStateChange(Update, tx.ObjectID, ContractDarcID, darcBuf),
-		}, nil, nil
-	} else if tx.Invoke.Command == "add" {
-		return nil, nil, errors.New("not implemented")
+	default:
+		return nil, nil, errors.New("Only invoke and spawn are defined yet")
 	}
-	return nil, nil, errors.New("invalid command: " + tx.Invoke.Command)
+}
+
+// ContractValue is a simple key/value storage where you
+// can put any data inside as wished.
+func (s *Service) ContractValue(cdb CollectionView, tx Instruction, c []Coin) ([]StateChange, []Coin, error) {
+	switch {
```

Comment:
There is no code here to validate that the instruction has exactly one of spawn, invoke, delete. Can be fixed later in a nicer way. See issue #1294 that I just filed.

Reply:
Having only one of `Spawn`, `Invoke`, `Delete` should be checked in omniledger itself.